### ISSUE TITLE
With upgrade from Python 3.9.4 to 3.9.5 DebugTraceTest.test_debug_trace and test_setting_new_trace_function started to fail

### ISSUE THREAD

**Issue by mcepl** [coveragepy/coveragepy#1161]
When running the packaging scripts for openSUSE package of python-coverage I get this failure in the test suite:
```
[   56s] + pytest-3.9 --ignore=_build.python36 --ignore=_build.python39 --ignore=_build.python38 -v -k 'not (test_get_encoded_zip_files or test_egg or test_doctest or test_unicode or test_version or test_multiprocessing_with_branching or test_farm or test_dothtml_not_python or test_one_of or test_bytes or test_encoding or test_multi or test_xdist_sys_path_nuttiness_is_fixed or test_debug_sys_ctracer)'
python 3.9.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3.9
[   56s] cachedir: .pytest_cache
[   56s] hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/abuild/rpmbuild/BUILD/coverage-5.5/.hypothesis/examples')
[   56s] rootdir: /home/abuild/rpmbuild/BUILD/coverage-5.5, configfile: setup.cfg
[   56s] plugins: forked-1.3.0, xdist-2.2.0, hypothesis-6.8.1, flaky-3.7.0
[   58s] collecting ... collected 807 items / 35 deselected / 772 selected
…
[   92s] _______________________ DebugTraceTest.test_debug_trace ________________________
[   92s]
[   92s] self = <tests.test_debug.DebugTraceTest testMethod=test_debug_trace>
[   92s]
[   92s]     def test_debug_trace(self):
[   92s]         out_lines = self.f1_debug_output(["trace"])
[   92s]
[   92s]         # We should have a line like "Tracing 'f1.py'"
[   92s] >       assert "Tracing 'f1.py'" in out_lines
[   92s] E       AssertionError: assert "Tracing 'f1.py'" in "Coverage code matching: <TreeMatcher ['/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage']>\nPython stdlib matching: ...of coverage.py\nNot tracing '/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/collector.py': is part of coverage.py\n"
[   92s]
[   92s] tests/test_debug.py:128: AssertionError
[   92s] _________________ GettraceTest.test_setting_new_trace_function _________________
[   92s]
[   92s] self = <tests.test_oddball.GettraceTest testMethod=test_setting_new_trace_function>
[   92s]
[   92s]     def test_setting_new_trace_function(self):
[   92s]         # https://github.com/nedbat/coveragepy/issues/436
[   92s]         self.check_coverage('''\
[   92s]             import sys
[   92s]
[   92s]             def tracer(frame, event, arg):
[   92s]                 print("%s: %s @ %d" % (event, frame.f_code.co_filename, frame.f_lineno))
[   92s]                 return tracer
[   92s]
[   92s]             def begin():
[   92s]                 sys.settrace(tracer)
[   92s]
[   92s]             def collect():
[   92s]                 t = sys.gettrace()
[   92s]                 assert t is tracer, t
[   92s]
[   92s]             def test_unsets_trace():
[   92s]                 begin()
[   92s]                 collect()
[   92s]
[   92s]             old = sys.gettrace()
[   92s]             test_unsets_trace()
[   92s]             sys.settrace(old)
[   92s]             a = 21
[   92s]             b = 22
[   92s]             ''',
[   92s]             lines=[1, 3, 4, 5, 7, 8, 10, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22],
[   92s]             missing="4-5, 11-12",
[   92s]         )
[   92s]
[   92s]         out = self.stdout().replace(self.last_module_name, "coverage_test")
[   92s]         expected = (
[   92s]             "call: coverage_test.py @ 10\n"
[   92s]             "line: coverage_test.py @ 11\n"
[   92s]             "line: coverage_test.py @ 12\n"
[   92s]             "return: coverage_test.py @ 12\n"
[   92s]         )
[   92s] >       assert expected == out
[   92s] E       AssertionError: assert 'call: covera...est.py @ 12\n' == 'call: /tmp/c...est.py @ 12\n'
[   92s] E         - call: /tmp/coverage_test/tests_test_oddball_GettraceTest_test_setting_new_trace_function_61005176/coverage_test.py @ 10
[   92s] E         - line: /tmp/coverage_test/tests_test_oddball_GettraceTest_test_setting_new_trace_function_61005176/coverage_test.py @ 11
[   92s] E         - line: /tmp/coverage_test/tests_test_oddball_GettraceTest_test_setting_new_trace_function_61005176/coverage_test.py @ 12
[   92s] E         - return: /tmp/coverage_test/tests_test_oddball_GettraceTest_test_setting_new_trace_function_61005176/coverage_test.py @ 12
[   92s] E         + call: coverage_test.py @ 10
[   92s] E         + line: coverage_test.py @ 11
[   92s] E         + line: coverage_test.py @ 12...
[   92s] E
[   92s] E         ...Full output truncated (2 lines hidden), use '-vv' to show
[   92s]
[   92s] tests/test_oddball.py:488: AssertionError
[   92s] =============================== warnings summary ===============================
[   92s] tests/test_data.py::CoverageDataTest::test_thread_stress
[   92s]   /usr/lib/python3.9/site-packages/_pytest/threadexception.py:75: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-32
[   92s]
[   92s]   Traceback (most recent call last):
[   92s]     File "/usr/lib64/python3.9/threading.py", line 954, in _bootstrap_inner
[   92s]       self.run()
[   92s]     File "/usr/lib64/python3.9/threading.py", line 892, in run
[   92s]       self._target(*self._args, **self._kwargs)
[   92s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/tests/test_data.py", line 494, in thread_main
[   92s]       covdata.add_lines(LINES_1)
[   92s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 438, in add_lines
[   92s]       self._choose_lines_or_arcs(lines=True)
[   92s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 495, in _choose_lines_or_arcs
[   92s]       with self._connect() as con:
[   92s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 300, in _connect
[   92s]       self._create_db()
[   92s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 250, in _create_db
[   92s]       db.executescript(SCHEMA)
[   92s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 1114, in executescript
[   92s]       self.con.executescript(script)
[   92s]   sqlite3.OperationalError: table coverage_schema already exists
[   92s]
[   92s]     warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))
[   92s]
[   92s] -- Docs: https://docs.pytest.org/en/stable/warnings.html
[   92s] ===Flaky Test Report===
[   92s]
[   92s] test_bug890 passed 1 out of the required 1 times. Success!
[   92s] test_for_leaks passed 1 out of the required 1 times. Success!
[   92s]
[   92s] ===End Flaky Test Report===
[   92s] =========================== short test summary info ============================
[   92s] FAILED tests/test_debug.py::DebugTraceTest::test_debug_trace - AssertionError...
[   92s] FAILED tests/test_oddball.py::GettraceTest::test_setting_new_trace_function
[   92s] ===== 2 failed, 764 passed, 6 skipped, 35 deselected, 1 warning in 36.18s ======
[   92s] Inefficient: MemoryFumblingTest ran 1 tests, 0 made files in a temp directory
[   92s] error: Bad exit status from /var/tmp/rpm-tmp.1hW18W (%check)
[   92s]
[   92s]
[   92s] RPM build errors:
[   92s]     Bad exit status from /var/tmp/rpm-tmp.1hW18W (%check)
[   92s]
[   92s] stitny failed "build python-coverage.spec" at Sat May  8 16:25:21 UTC 2021.
[   92s]

The buildroot was: /var/tmp/build-root/standard-x86_64
python-coverage@stitny$
```

**To Reproduce**
How can we reproduce the problem?  Please *be specific*. Don't just link to a failing CI job. Answer the questions below:
1. What version of Python are you using? 3.9.5
2. What version of coverage.py are you using? The output of `coverage debug sys` is helpful. 5.5
```
[    6s] + python3.9 -mcoverage debug sys
[    6s] -- sys -------------------------------------------------------
[    6s]                         version: 5.5
[    6s]                        coverage: /home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/__init__.py
[    6s]                          tracer: -none-
[    6s]                         CTracer: unavailable
[    6s]            plugins.file_tracers: -none-
[    6s]             plugins.configurers: -none-
[    6s]       plugins.context_switchers: -none-
[    6s]               configs_attempted: .coveragerc
[    6s]                                  setup.cfg
[    6s]                                  tox.ini
[    6s]                                  pyproject.toml
[    6s]                    configs_read: /home/abuild/rpmbuild/BUILD/coverage-5.5/setup.cfg
[    6s]                                  /home/abuild/rpmbuild/BUILD/coverage-5.5/tox.ini
[    6s]                     config_file: None
[    6s]                 config_contents: -none-
[    6s]                       data_file: -none-
[    6s]                          python: 3.9.5 (default, May 05 2021, 15:23:13) [GCC]
[    6s]                        platform: Linux-5.12.0-2-default-x86_64-with-glibc2.33
[    6s]                  implementation: CPython
[    6s]                      executable: /usr/bin/python3.9
[    6s]                    def_encoding: utf-8
[    6s]                     fs_encoding: utf-8
[    6s]                             pid: 23107
[    6s]                             cwd: /home/abuild/rpmbuild/BUILD/coverage-5.5
[    6s]                            path: /home/abuild/rpmbuild/BUILD/coverage-5.5
[    6s]                                  /usr/lib64/python39.zip
[    6s]                                  /usr/lib64/python3.9
[    6s]                                  /usr/lib64/python3.9/lib-dynload
[    6s]                                  /usr/lib64/python3.9/site-packages
[    6s]                                  /usr/lib64/python3.9/_import_failed
[    6s]                                  /usr/lib/python3.9/site-packages
[    6s]                     environment: PYTHONHASHSEED = 0
[    6s]                    command_line: /home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/__main__.py debug sys
[    6s]                 sqlite3_version: 2.6.0
[    6s]          sqlite3_sqlite_version: 3.35.5
[    6s]              sqlite3_temp_store: 0
[    6s]         sqlite3_compile_options: COMPILER=gcc-10.3.0
[    6s]                                  ENABLE_API_ARMOR
[    6s]                                  ENABLE_COLUMN_METADATA
[    6s]                                  ENABLE_DBSTAT_VTAB
[    6s]                                  ENABLE_FTS3
[    6s]                                  ENABLE_FTS4
[    6s]                                  ENABLE_FTS5
[    6s]                                  ENABLE_HIDDEN_COLUMNS
[    6s]                                  ENABLE_JSON1
[    6s]                                  ENABLE_MATH_FUNCTIONS
[    6s]                                  ENABLE_RBU
[    6s]                                  ENABLE_RTREE
[    6s]                                  ENABLE_UNLOCK_NOTIFY
[    6s]                                  ENABLE_UPDATE_DELETE_LIMIT
[    6s]                                  HAVE_ISNAN
[    6s]                                  SECURE_DELETE
[    6s]                                  SOUNDEX
[    6s]                                  TEMP_STORE=1
[    6s]                                  THREADSAFE=1
```
1. What versions of what packages do you have installed? The output of `pip freeze` is helpful.
[Complete build log](https://github.com/nedbat/coveragepy/files/6446368/_log.txt)

1. What code are you running?  Give us a specific commit of a specific repo that we can check out.

see in the log

**Expected behavior**
Testsuite to pass

**Additional context**
It is failing with the errors above.

**Comment by nedbat** [comment]
This fix was already underway for 3.10.0b1. I didn't realize they would make a change like this between 3.9.4 and 3.9.5....

**Comment by mcepl** [comment]
Hmm, strange. Python 3.6 and 3.9 pass, but 3.8 fails:
```
[  227s] _____________________ CoverageDataTest.test_thread_stress ______________________
[  227s] 
[  227s] self = <tests.test_data.CoverageDataTest testMethod=test_thread_stress>
[  227s] 
[  227s]     def test_thread_stress(self):
[  227s]         covdata = CoverageData()
[  227s]     
[  227s]         def thread_main():
[  227s]             """Every thread will try to add the same data."""
[  227s]             covdata.add_lines(LINES_1)
[  227s]     
[  227s]         threads = [threading.Thread(target=thread_main) for _ in range(10)]
[  227s]         for t in threads:
[  227s]             t.start()
[  227s]         for t in threads:
[  227s]             t.join()
[  227s]     
[  227s] >       self.assert_lines1_data(covdata)
[  227s] 
[  227s] tests/test_data.py:502: 
[  227s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  227s] tests/test_data.py:90: in assert_lines1_data
[  227s]     self.assert_line_counts(covdata, SUMMARY_1)
[  227s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  227s] 
[  227s] self = <tests.test_data.CoverageDataTest testMethod=test_thread_stress>
[  227s] covdata = <CoverageData @0x7fed144da340 _no_disk=False _basename='/tmp/coverage_test/tests_test_data_CoverageDataTest_test_threa...58 _have_used=True _has_lines=True _has_arcs=False _current_context=None _current_context_id=1 _query_context_ids=None>
[  227s] counts = {'a.py': 2, 'b.py': 1}, fullpath = False
[  227s] 
[  227s]     def assert_line_counts(self, covdata, counts, fullpath=False):
[  227s]         """Check that the line_counts of `covdata` is `counts`."""
[  227s] >       assert line_counts(covdata, fullpath) == counts
[  227s] E       AssertionError: assert {'a.py': 3, 'b.py': 3} == {'a.py': 2, 'b.py': 1}
[  227s] E         Differing items:
[  227s] E         {'a.py': 3} != {'a.py': 2}
[  227s] E         {'b.py': 3} != {'b.py': 1}
[  227s] E         Full diff:
[  227s] E         - {'a.py': 2, 'b.py': 1}
[  227s] E         ?          ^          ^
[  227s] E         + {'a.py': 3, 'b.py': 3}...
[  227s] E         
[  227s] E         ...Full output truncated (2 lines hidden), use '-vv' to show
[  227s] 
[  227s] tests/test_data.py:82: AssertionError
[  227s] =============================== warnings summary ===============================
[  227s] tests/test_data.py::CoverageDataTest::test_thread_stress
[  227s]   /usr/lib/python3.8/site-packages/_pytest/threadexception.py:75: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-41
[  227s]   
[  227s]   Traceback (most recent call last):
[  227s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 1060, in execute
[  227s]       return self.con.execute(sql, parameters)
[  227s]   sqlite3.OperationalError: no such table: line_bits
[  227s]   
[  227s]   During handling of the above exception, another exception occurred:
[  227s]   
[  227s]   Traceback (most recent call last):
[  227s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 1065, in execute
[  227s]       return self.con.execute(sql, parameters)
[  227s]   sqlite3.OperationalError: no such table: line_bits
[  227s]   
[  227s]   During handling of the above exception, another exception occurred:
[  227s]   
[  227s]   Traceback (most recent call last):
[  227s]     File "/usr/lib64/python3.8/threading.py", line 932, in _bootstrap_inner
[  227s]       self.run()
[  227s]     File "/usr/lib64/python3.8/threading.py", line 870, in run
[  227s]       self._target(*self._args, **self._kwargs)
[  227s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/tests/test_data.py", line 494, in thread_main
[  227s]       covdata.add_lines(LINES_1)
[  227s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 447, in add_lines
[  227s]       existing = list(con.execute(query, (file_id, self._current_context_id)))
[  227s]     File "/home/abuild/rpmbuild/BUILD/coverage-5.5/coverage/sqldata.py", line 1082, in execute
[  227s]       raise CoverageException("Couldn't use data file {!r}: {}".format(self.filename, msg))
[  227s]   coverage.misc.CoverageException: Couldn't use data file '/tmp/coverage_test/tests_test_data_CoverageDataTest_test_thread_stress_74715056/.coverage': no such table: line_bits
[  227s]   
[  227s]     warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))
[  227s] 
[  227s] -- Docs: https://docs.pytest.org/en/stable/warnings.html
[  227s] ===Flaky Test Report===
[  227s] 
[  227s] test_bug890 passed 1 out of the required 1 times. Success!
[  227s] test_for_leaks passed 1 out of the required 1 times. Success!
[  227s] 
[  227s] ===End Flaky Test Report===
[  227s] =========================== short test summary info ============================
[  227s] FAILED tests/test_data.py::CoverageDataTest::test_thread_stress - AssertionEr...
[  227s] ===== 1 failed, 765 passed, 6 skipped, 35 deselected, 1 warning in 58.42s ======
```
[Complete build log](https://github.com/nedbat/coveragepy/files/6448606/_log.txt)

**Comment by nedbat** [comment]
Ugh, yes I see it now too on 3.8.10.  Was this really a necessary change in an x.x.10 release? :(   Perhaps the coverage.py tests are too detailed....

**Comment by mcepl** [comment]
> Ugh, yes I see it now too on 3.8.10. Was this really a necessary change in an x.x.10 release? :( Perhaps the coverage.py tests are too detailed....

I don't argue with you … I am a maintainer of of couple of thousand of Python packages for SUSE, so I **HATE** any API instability.

**Comment by nedbat** [comment]
OK, now 3.8.10 is fixed in 01cbb875.
